proposals/0000-exception-backtraces.rst

```diff
@@ -0,0 +1,513 @@
+relevant metadata about the exception.
+
+As a first step towards attaching backtraces to exceptions, we propose to extend
+``SomeException`` to carry an additional "context" value, consisting of a collection
+of dynamically-typed annotations. The dynamic nature of these annotations
+allows the mechanism to handle not just backtraces but any arbitrary
+context-relevant value which may be inspected by a handler for analysis and logging.
+To avoid breaking existing users of ``SomeException``, we introduce this
+context as an implicit parameter constraint:  ::
+
+    data SomeException where
+        SomeException :: forall a. (Exception a, ?exceptionContext :: ExceptionContext)
+                      => a -> SomeException
+
+    data ExceptionContext = ExceptionContext [SomeExceptionAnnotation]
+
+    -- | Contexts are merged by concatenation
+    instance Semigroup ExceptionContext
+    instance Monoid ExceptionContext
+
+Following the example of the ``Exception`` class, we propose that exception
+annotations implement a minimal typeclass which provides ``Typeable`` evidence
+and the ability to render the annotation to a ``String``: ::
+
+    data SomeExceptionAnnotation where
+        SomeExceptionAnnotation :: forall a. (ExceptionAnnotation a)
+                                => a -> SomeExceptionAnnotation
+
+    class Typeable a => ExceptionAnnotation a where
+        displayExceptionAnnotation :: a -> String
+        
+        default displayExceptionAnnotation :: (Show a) => a -> String
+        displayExceptionAnnotation = show
+
+To allow users to populate this new annotation field we propose that the
+``Exception`` class be extended with a new ``toExceptionWithContext`` method: ::
+
+    class (Typeable e, Show e) => Exception e where
+        -- These are unchanged:
+        toException            :: e -> SomeException
+        fromException          :: SomeException -> Maybe e
+        displayException       :: e -> String
+
+        -- This is new:
+        toExceptionWithContext :: e -> ExceptionContext -> SomeException
+        toExceptionWithContext e ?exceptionContext = SomeException e
+
+        -- toException is implemented in terms of toExceptionWithContext
+        toException e = toExceptionWithContext e mempty
+
+The ``Exception`` instance for ``SomeException`` would accumulate and display
+contexts: ::
+
+    instance Exception SomeException where
+        -- This is unchanged:
+        toException se = se
+        fromException = Just
+
+        -- toExceptionWithContext *adds* context to an existing SomeException:
+        toExceptionWithContext se@(SomeException e) ctxt =
+            SomeException e
+          where ?exc_context = ctxt <> exceptionContext se
+
+        -- displayException shows context after the exception itself:
+        displayException (SomeException e) =
+            displayException e ++ "\n" ++ displayExceptionContext ?exc_context
```

Comment:
Yes, this is true. However, I can see few good ways of addressing this. One way would be to add yet another method to `Exception`:
```
class Exception a where
    ...
    displayException :: a -> String
    displayException = show

    displayExceptionWithContext :: a -> String
    displayExceptionWithContext x = displayException x ++ "\n" ++ displayExceptionContext ?exc_context
```
`displayExceptionWithContext` would then become the preferred function for formatting exceptions for the user. However, this merely moves the problem: instead of needing to update all *definition* sites of `displayException` you now rather need to update all call-sites. A rather [unscientific poll of Hackage](https://hackage-search.serokell.io/?q=displayException) suggests that call-sites may actually be more numerous than method definitions, so I suspect that this change would actually be worse than the current proposal.

Reply:
Where is `?exc_context` brought into scope here? I think, without a `SomeException` wrapper, we just won't have context. Which isn't necessarily a problem, *generally* if you have a `SomeException` it's because you don't know more about what you want to be catching.

Reply:
Yes, quite right. I don't know what I was thinking when I wrote this.